File: AlgorithmicToolbox/MaxPairwiseProduct.cpp
// Problem description: Find the maximum product of two distinct numbers in a
// sequence of non-negative integers.

#include <iostream>
#include <vector>
using namespace std;

int MaxPairwiseProduct(vector<int> arr, long long n)
{
    int a = 0, b = 0;
    for (long long i = 0; i < n; i++)
    {
        if (arr[i] > a)
        {
            b = a;
            a = arr[i];
            continue;
        }
        if (arr[i] > b)
        {
            b = arr[i];
        }
    }
    return (a * b);
}

int main()
{
    long long n;
    cin >> n;
    vector<int> arr;
    for (long long i = 0; i < n; i++)
    {
        int temp;
        cin >> temp;
        arr.push_back(temp);
    }
    cout << MaxPairwiseProduct(arr, n);
    return 0;
}

File: AlgorithmicToolbox/MaxPairwiseProduct.py
# Problem description: Find the maximum product of two distinct numbers in a
# sequence of non-negative integers.

import doctest


def MaxPairwiseProduct(integers, n):
    """Return the maximum pair product from the list of integers of length n.

    Examples:
    >>> MaxPairwiseProduct([1,2,3], 3)
    6
    >>> MaxPairwiseProduct([7,5,14,2,8,8,10,1,2,3], 10)
    140
    """

    a, b = 0, 0
    for integer in integers:
        if(integer > a):
            b = a
            a = integer
            continue
        if(integer > b):
            b = integer
    return(a*b)


if __name__ == "__main__":
    doctest.testmod()
    n = int(input())
    integers = [int(i) for i in input().split()]
    print(MaxPairwiseProduct(integers, n))
